pipelines/toast_ground_sim.py

```diff
@@ -1310,6 +1339,94 @@ def main():
 
+    local_dets = extract_local_dets(data)
+    print(local_dets)
+
+    run_pysm = True
+    if run_pysm:
+
+        # Simulate input sky
+
+        dist_rings = tm.DistRings(comm.comm_world,
+                            nside = args.nside,
+                            nnz = 3)
+
+        pysm_sky_components = [
+            'synchrotron',
+            'dust',
+            'freefree',
+            'cmb',
+            'ame',
+        ]
+        pysm_sky_config = dict()
+        for component_model in args.input_pysm_model.split(','):
+            full_component_name = \
+                    [each for each in pysm_sky_components
+                     if each.startswith(component_model[0])][0]
+            pysm_sky_config[full_component_name] = component_model
+
+        bandpasses = {}
+        fwhm_deg = {}
+        N_POINTS_BANDPASS = 10  # possibly take as parameter
+        for det in local_dets:
+            bandcenter, bandwidth, fwhm_deg[det] = extract_detector_parameters(det, schedules)
+            bandpasses[det] = \
+                (np.linspace(bandcenter-bandwidth/2,
+                             bandcenter+bandwidth/2,
+                             N_POINTS_BANDPASS),
+                 np.ones(N_POINTS_BANDPASS))
+        pysm_sky = tm.PySMSky(local_pixels=dist_rings.local_pixels,
+                              nside=args.nside,
+                              pysm_sky_config=pysm_sky_config,
+                              bandpasses=bandpasses)
+        local_maps = dict()  # FIXME use Cache instead
+        pysm_sky.exec(local_maps, out="sky")
+
+        lmax = 512
+
+        npix = hp.nside2npix(args.nside)
+        distmap = tm.DistPixels(
+            comm=comm.comm_world, size=npix, nnz=3,
+            dtype=np.float32, submap=subnpix, local=localsm)
+
+        if comm.comm_world.rank == 0:
+            print('Collecting, Broadcasting map', flush=args.flush)
+        start = MPI.Wtime()
+
+        full_map_rank0 = {}
+        for det, fwhm in fwhm_deg.items():
+            # LibSharp also supports transforming multiple channels together each with own beam
+            smooth = tm.LibSharpSmooth(comm.comm_world, signal_map="sky_"+det,
+                                       lmax=lmax, grid=dist_rings.libsharp_grid,
+                                       fwhm_deg=fwhm, beam=None)
+            smooth.exec(local_maps)
+
+            n_components = 3
+
+            full_map_rank0 = assemble_map_on_rank0(comm, local_maps["sky_"+det], n_components, npix)
+            #full_map_rank0 dict contains on rank 0 the smoothed PySM maps
+
+            if comm.comm_world.rank == 0:
+                print(full_map_rank0[:10])
+
+            distmap.broadcast_healpix_map(full_map_rank0)
+            scansim = tt.OpSimScan(distmap=distmap, out='signal_'+det)
+            scansim.exec(data)
+            if comm.comm_world.rank == 0:
+                print("Time domain signal", data["signal_fake_0A"][:10])
```

Comment:
@tskisner @keskitalo  at this point I have `signal_detname` inside `data`, in the orginal script there was only one single signal, do you have any suggestion on how to handle this? I don't understand what `sigcopy` is doing after this point.

Reply:
Ah, now I see.  You are using the OpSimScan operator to do the actual timestream generation.  Originally (before the decision to reduce the data to a single process), I had thought that the sampling would be part of the "OpSimPySM" operator.  But since you are using the OpSimScan operator to do the sampling, then all the pysm work becomes just map-domain classes and functions.  In the medium term, it might still be nice to put all this into its own operator class (tiny amount of copied code from OpSimScan), just so that all the "setup" is in one place and we can then call it more easily from other pipelines.

Reply:
sure, I'll wrap it in an operator when it is almost completed, it is easier to develop directly in the script